x/mongo/driver/topology/server.go

```diff
@@ -222,23 +224,19 @@ func (s *Server) Connection(ctx context.Context) (driver.Connection, error) {
 		return nil, ErrServerClosed
 	}
 
-	conn, err := s.pool.get(ctx)
+	poolGen := s.pool.getGeneration()
+	connImpl, err := s.pool.get(ctx)
 	if err != nil {
-		wrappedConnErr := unwrapConnectionError(err)
-		if wrappedConnErr == nil {
-			return nil, err
+		// ProcessHandshakeError needs this to check the staleness of the error
+		connImpl = &connection{
+			pool:       s.pool,
+			generation: poolGen,
 		}
-
-		// Since the only kind of ConnectionError we receive from pool.Get will be an initialization
-		// error, we should set the description.Server appropriately.
-		desc := description.NewServerFromError(s.address, wrappedConnErr)
-		s.updateDescription(desc)
-		s.pool.clear()
-
+		s.ProcessHandshakeError(err, &Connection{connection: connImpl})
```

Comment:
Creating a fake connection here seems weird to me. Can we have the function definition for `ProcessHandshakeError` be `ProcessHandshakeError(err, startingGenerationNumber int)` and then it can internally access the pool's current generation number to see if the error is stale?